streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -471,6 +483,76 @@ trait ZStream[-R, +E, +A] extends Serializable { self =>
       ZStream.fromQueue(queue).unTake
     }
 
+  /**
+   * Switches over to the stream produced by the provided function in case this one
+   * fails with a typed error.
+   */
+  final def catchAll[R1 <: R, E2, A1 >: A](f: E => ZStream[R1, E2, A1]): ZStream[R1, E2, A1] =
+    self.catchAllCause(_.failureOrCause.fold(f, ZStream.halt))
+
+  /**
+   * Switches over to the stream produced by the provided function in case this one
+   * fails. Allows recovery from all errors, except external interruption.
+   */
+  final def catchAllCause[R1 <: R, E2, A1 >: A](f: Cause[E] => ZStream[R1, E2, A1]): ZStream[R1, E2, A1] = {
+    sealed abstract class State
+    object State {
+      case object NotStarted extends State
+      case object Self       extends State
+      case object Other      extends State
+    }
+
+    ZStream.fromInputStreamManaged {
+      for {
+        finalizer        <- Ref.make[Exit[_, _] => ZIO[R1, Nothing, Any]](_ => UIO.unit).toManaged_
+        _                <- ZManaged.finalizerExit(e => finalizer.get.flatMap(_.apply(e)))
+        selfInputStream  <- Ref.make[InputStream[R, E, A]](InputStream.end).toManaged_
```

Comment:
This pattern occurs so pervasively I think we should write a combinator to handle it (`finalizer`?). Basically creating a `Ref`, with some default action, then running some code, updating the default, and finally ensuring the finalizer is run at the end. Could be reasonably nice to use with `ZManaged`, I think.

Reply:
Good idea. This looked like a combinator for `ZManaged` - `ZManaged.finalizerRef`.